src/tiled/objectsdock.cpp

```diff
@@ -221,47 +233,90 @@ void ObjectsDock::objectProperties()
     emit mMapDocument->editCurrentObject();
 }
 
-void ObjectsDock::saveExpandedGroups()
+QModelIndex ObjectsDock::getGroupIndex(ObjectGroup *og)
 {
-    mExpandedGroups[mMapDocument].clear();
-
     const auto proxyModel = static_cast<QAbstractProxyModel*>(mObjectsView->model());
+    const QModelIndex sourceIndex = mMapDocument->mapObjectModel()->index(og);
+    return proxyModel->mapFromSource(mObjectsView->proxyModel()->mapFromSource(sourceIndex));
+}
+
+void ObjectsDock::saveExpandedGroups(QMap<MapDocument*, QList<ObjectGroup*> > &expansionState)
+{
+    expansionState[mMapDocument].clear();
+
     const auto &objectGroups = mMapDocument->map()->objectGroups();
 
-    for (ObjectGroup *og : objectGroups) {
-        const QModelIndex sourceIndex = mMapDocument->mapObjectModel()->index(og);
-        const QModelIndex index = proxyModel->mapFromSource(sourceIndex);
-        if (mObjectsView->isExpanded(index))
-            mExpandedGroups[mMapDocument].append(og);
-    }
+    for (ObjectGroup *og : objectGroups)
+        if (mObjectsView->isExpanded(getGroupIndex(og)))
+            expansionState[mMapDocument].append(og);
 }
 
-void ObjectsDock::restoreExpandedGroups()
+void ObjectsDock::restoreExpandedGroups(QMap<MapDocument*, QList<ObjectGroup*> > &expansionState)
 {
-    const auto objectGroups = mExpandedGroups.take(mMapDocument);
-    for (ObjectGroup *og : objectGroups) {
-        const QModelIndex sourceIndex = mMapDocument->mapObjectModel()->index(og);
-        const QModelIndex index = static_cast<QAbstractProxyModel*>(mObjectsView->model())->mapFromSource(sourceIndex);
-        mObjectsView->setExpanded(index, true);
-    }
+    const auto objectGroups = expansionState.take(mMapDocument);
+
+    for (ObjectGroup *og : objectGroups)
+        mObjectsView->setExpanded(getGroupIndex(og), true);
 }
 
 void ObjectsDock::documentAboutToClose(Document *document)
 {
-    if (MapDocument *mapDocument = qobject_cast<MapDocument*>(document))
+    if (MapDocument *mapDocument = qobject_cast<MapDocument*>(document)) {
+        mFilterStrings.remove(mapDocument);
         mExpandedGroups.remove(mapDocument);
+    }
+}
+
+void ObjectsDock::filterObjects()
+{
+    // Save the prefilter expansion state if this is the first applied filter
+    if(mFilterWasEmpty[mMapDocument]) {
+        mFilterWasEmpty[mMapDocument] = false;
```

Comment:
I tried doing that but couldn't get it to work properly.

The filter is shared between documents, so suppose we switched from a non-filtered map to a filtered map, the above condition will be satisfied and the state of the new map will be saved overwriting the original saved state, I couldn't figure a clean way to prevent this from happening.

I changed the mFilterWasEmpty map to a single bool variable and got it work properly but had to introduce a small hack for this particular case.